### PR TITLE
Precommit client

### PR DESCRIPTION
**Pull request by dpagano**
First version of precommit client

### CODE REVIEW

File: teamscale_client/client.py [hunk 1]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
         with open(file_path) as json_file:
             json_data = json.load(json_file)
             return json_data
```
Comment (nihathrael) [inline teamscale_client/client.py:521]: Multiple empty lines between methods, reduce to one?

File: teamscale_client/client.py [hunk 2]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
...
+
+        added_findings = self._findings_from_json(response.json()['addedFindings'])
+        findings_in_changed_code = self._findings_from_json(response.json()['findingsInChangedCode'])
+        removed_findings = self._findings_from_json(response.json()['removedFindings'])
+
+        return added_findings, removed_findings, findings_in_changed_code
+
+
+    def _findings_from_json(self, findings_json):
+        """Parses JSON encoded findings."""
```
Comment (nihathrael) [inline teamscale_client/client.py]: Add return type documentation?

File: teamscale_client/client.py [hunk 3]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
...
+
+
+    def get_precommit_analysis_results(self):
+        """Gets precommit analysis results."""
+        service_url = self.get_project_service_url("pre-commit")
+
+        while True:
+            response = self.get(service_url)
+            if response.json() is None:
+                time.sleep(2)
```
Comment (nihathrael) [inline teamscale_client/client.py:545]: Aren't we only allowed to do this every 5 seconds?

File: teamscale_client/client.py [hunk 4]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
...
+
+    def upload_files_for_precommit_analysis(self, timestamp, precommit_data):
+        """Uploads the provided files for precommit analysis."""
+        service_url = self.get_project_service_url("pre-commit") + self._get_timestamp_parameter(timestamp)
+
+        return self.put(service_url, data=to_json(precommit_data))
+
+
+    def get_precommit_analysis_results(self):
+        """Gets precommit analysis results."""
```
Comment (nihathrael) [inline teamscale_client/client.py]: Document return type?

File: teamscale_client/client.py [hunk 5]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
         with open(file_path) as json_file:
             json_data = json.load(json_file)
             return json_data
+
+
+    def upload_files_for_precommit_analysis(self, timestamp, precommit_data):
+        """Uploads the provided files for precommit analysis."""
```
Comment (nihathrael) [inline teamscale_client/client.py]: Document parameters?

File: teamscale_client/client.py [hunk 6]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
...
+        """Parses JSON encoded findings."""
+        return [Finding(finding_type_id=x['typeId'], message=x['message'],
+                        assessment=x['assessment'], start_offset=x['location']['rawStartOffset'],
+                        end_offset=x['location']['rawEndOffset'], start_line=x['location']['rawStartLine'],
+                        end_line=x['location']['rawEndLine'], uniform_path=x['location']['uniformPath'])
+                for x in findings_json]
+
+
+    def get_findings(self, uniform_path, timestamp):
+        """Retrieves the list of findings in the currently active project for the given uniform path
```
Comment (nihathrael) [inline teamscale_client/client.py]: Exactly for this path, or recursively as well? Please document

File: teamscale_client/client.py [hunk 7]
```diff
@@ -518,3 +518,68 @@ def read_json_from_file(cls, file_path):
...
+            ServiceError: If anything goes wrong
+        """
+        service_url = self.get_project_service_url("findings") + uniform_path
+        parameters = {
+            "t": self._get_timestamp_parameter(timestamp=timestamp),
+            "recursive": True,
+            "all": True
+        }
+        headers = {'Accept': 'application/json'}
+        response = requests.get(service_url, params=parameters, auth=self.auth_header, verify=self.sslverify,
```
Comment (nihathrael) [inline teamscale_client/client.py]: Use self.get()?

File: teamscale_client/git_utils.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from __future__ import print_function
+
+import os
+from pygit2 import Repository, GIT_STATUS_WT_DELETED, GIT_STATUS_WT_MODIFIED, GIT_STATUS_WT_RENAMED, \
+    GIT_STATUS_WT_TYPECHANGE, GIT_STATUS_INDEX_NEW, GIT_STATUS_INDEX_MODIFIED, GIT_STATUS_INDEX_DELETED, \
+    discover_repository
+
```
Comment (nihathrael) [inline teamscale_client/git_utils.py:9]: Lots of multiple empty lines again

File: teamscale_client/git_utils.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        Returns:
+            str: The timestamp of the last commit in the provided repository.
+    """
+    repo = Repository(path_to_repository)
+    head = repo.revparse_single('HEAD')
+    return head.commit_time
+
+
+def filter_changed_files(changed_files):
```
Comment (nihathrael) [inline teamscale_client/git_utils.py]: Do we need this?

File: teamscale_client/git_utils.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        Args:
+            path_to_repository (str): Path to the Git repository
+
+        Returns:
+            List(str): List of filenames of all changed files in the provided repository.
+    """
+    repo = Repository(path_to_repository)
+    status_entries = repo.status()
+
+    stati_considered_for_precommit = GIT_STATUS_INDEX_NEW | \
```
Comment (nihathrael) [inline teamscale_client/git_utils.py]: Move to constant?

File: teamscale_client/git_utils.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        Args:
+            path_to_repository (str): Path to the Git repository
+
+        Returns:
+            List(str): List of filenames of all deleted files in the provided repository.
+    """
+    repo = Repository(path_to_repository)
+    status_entries = repo.status()
+
+    stati_deleted = GIT_STATUS_INDEX_DELETED | \
```
Comment (nihathrael) [inline teamscale_client/git_utils.py]: Constant?

File: teamscale_client/precommit_client.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
+from __future__ import absolute_import
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py:1]: Multiple empty lines

File: teamscale_client/precommit_client.py [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+
+    def upload_precommit_data(self):
+        """Uploads the currently changed files for precommit analysis."""
+        current_branch = get_current_branch(self.repository_path)
+        self.teamscale_client.branch = current_branch
+        print('Current branch: %s' % current_branch)
+
+        parent_commit_timestamp = get_current_timestamp(self.repository_path)
+        print('Parent commit timestamp: %s' % parent_commit_timestamp)
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: Move timestamp and branch on same line? Do we really need the timestamp?

File: teamscale_client/precommit_client.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+        changed_files = get_changed_files_and_content(self.repository_path)
+        deleted_files = get_deleted_files(self.repository_path)
+
+        precommit_data = PreCommitUploadData(uniformPathToContentMap=changed_files, deletedUniformPaths=deleted_files)
+        self.teamscale_client.upload_files_for_precommit_analysis(
+            datetime.datetime.fromtimestamp(int(parent_commit_timestamp)), precommit_data)
+
+
+    def get_precommit_results(self):
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: Rename to wait_and_get_precommit_result()

File: teamscale_client/precommit_client.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    def _format_findings(self, findings):
+        """Formats the given findings as error or warning strings."""
+        if len(findings) == 0:
+            return ['> No findings.']
+        sorted_findings = sorted(findings)
+        return [os.path.join(self.repository_path, finding.uniformPath) + ':' + unicode(finding.startLine) + ':0: ' +
+                self._severity_string(finding=finding) + ': ' + finding.message for finding in sorted_findings]
+
+
+    def _severity_string(self, finding):
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: `return 'error' if finding.assessment == 'RED' else 'warning'`?

File: teamscale_client/precommit_client.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        """Formats the given finding's assessment as severity."""
+        if finding.assessment == 'RED':
+            return 'error'
+        else:
+            return 'warning'
+
+
+precommit_config_filename = '.teamscale-precommit.config'
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: Move to the top?

File: teamscale_client/precommit_client.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    """Reads the precommit analysis configuration and creates a precommit client with the corresponding config."""
+    parser = ConfigParser()
+    parser.read(config_file)
+    config = TeamscaleConfig(url=parser.get('teamscale', 'url'), username=parser.get('teamscale', 'username'),
+                             access_token=parser.get('teamscale', 'access_token'),
+                             project_id=parser.get('project', 'id'))
+    return PrecommitClient(teamscale_config=config, repository_path=repo_path, analyzed_file=parsed_args.path[0])
+
+
+def analyze():
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: rename to `run()` or `execute()`?

File: teamscale_client/precommit_client.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                        help='When this option is set, all existing findings in the repo are fetched in addition '
+                             'to precommit findings. (default: False)')
+    return parser.parse_args()
+
+
+def configure_precommit_client(config_file, repo_path, parsed_args):
+    """Reads the precommit analysis configuration and creates a precommit client with the corresponding config."""
+    parser = ConfigParser()
+    parser.read(config_file)
+    config = TeamscaleConfig(url=parser.get('teamscale', 'url'), username=parser.get('teamscale', 'username'),
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: Move the parsing into the Teamscale Config?

File: teamscale_client/precommit_client.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+    config_file = os.path.join(repo_path, precommit_config_filename)
+    if not os.path.exists(config_file) or not os.path.isfile(config_file):
+        raise RuntimeError('Config file could not be found: %s' % config_file)
+    precommit_client = configure_precommit_client(config_file=config_file, repo_path=repo_path, parsed_args=parsed_args)
+
+    print('Uploading changes for precommit analysis for repo at %s...' % repo_path)
+    precommit_client.upload_precommit_data()
+
+    # We need to wait for the analysis to pick up the new code otherwise we get old findings
+    time.sleep(2)
```
Comment (nihathrael) [inline teamscale_client/precommit_client.py]: We might not need this in the future, I think this was fixed as part of a bug recently

File: config/teamscale-precommit-vscode-task.json [hunk 20]
```diff
@@ -0,0 +1,27 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=733558
+    // for the documentation about the tasks.json format
+    "version": "2.0.0",
+    "tasks": [
+        {
+            "label": "Teamscale Precommit Analysis",
+            "type": "shell",
+            "command": "python -c teamscale_client.precommit_client import analyze;analyze() ${file}",
```
Comment (nihathrael) [inline config/teamscale-precommit-vscode-task.json]: I think -c needs " around the actual snippet that is supposed to be executed

File: config/teamscale-precommit-vscode-task.json [hunk 21]
```diff
@@ -0,0 +1,27 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=733558
```
Comment (nihathrael) [inline config/teamscale-precommit-vscode-task.json]: Remove comments? Not legal in JSON

File: teamscale_client/client.py (hunk 3 above)
Comment (dpagano) [reply teamscale_client/client.py]: Only for PUT

File: teamscale_client/git_utils.py (hunk 8 above)
Comment (dpagano) [reply teamscale_client/git_utils.py:9]: 2 is standard PEP8. But let's make it consistent for now and bulk change it later on.

File: teamscale_client/git_utils.py (hunk 9 above)
Comment (dpagano) [reply teamscale_client/git_utils.py]: Not yet, but I've changed it so we need it :-)

File: teamscale_client/precommit_client.py (hunk 13 above)
Comment (dpagano) [reply teamscale_client/precommit_client.py]: Don't understand

File: teamscale_client/precommit_client.py (hunk 19 above)
Comment (dpagano) [reply teamscale_client/precommit_client.py]: I'll add a comment.

**Pull request merged**